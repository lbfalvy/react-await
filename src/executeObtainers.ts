
import { divide, unzip, zip } from '@lbfalvy/array-utils'
import { useCachedMap } from './useCachedMap'
import React from 'react'


export type Obtainer <T> = 
    ( () => T ) | 
    [ () => T , ... any [] ]

type ObtainerPair = [ string , Obtainer<any> ]
type AnyPair = [ string , any ]


const { length : nameStart } = 'obtain';


/**
 *  @brief Picks out all the obtainers and executes 
 *  them unless they were also provided last round.
 */

export function executeObtainers (
    entries : AnyPair []
) : [ AnyPair [] , () => void ] {
    
    const [ obtained , provided ] = 
        divide<ObtainerPair,AnyPair>(entries,isObtainerPair);
    
    const [ names , obtainers ] = 
        unzip(obtained);
    
    const functions = obtainers
        .map(updateObtainer);
    
    const [ results , reload ] = 
        useCachedMap(functions,f => f());
    
    const resultNames = names
        .map(formatName);
            
    const zipped = [
        ... provided ,
        ... zip(resultNames,results)
    ]
            
    return [ zipped , reload ]
}


function updateObtainer ( obtainer : Obtainer ) : Function<any> {
    
    if(typeof obtainer == 'function')
        return React.useCallback(obtainer,[ obtainer ])
    
    const [ f , ... dependencies ] = obtainer;
    
    // The instance must change if the dependency array does
    
    return React.useCallback(() => f(),dependencies)
}


function formatName ( name : string ) : string {
    return name[nameStart].toLowerCase() 
         + name.slice(nameStart + 1)
}


function isObtainerPair ( pair ) : pair is ObtainerPair {
    
    const [ name , value ] = pair;
    
    // Named obtain<Prop>
    
    if(! name.startsWith('obtain') )
        return false;
    
    // <Prop> is not empty
    
    if(name === 'obtain')
        return false;
        
    // Their value is either
    
    if(typeof value == 'function')
        return true;
        
    if(
        Array.isArray(value) &&
        typeof value[0] == 'function'
    ) return true;
    
    return false;
}