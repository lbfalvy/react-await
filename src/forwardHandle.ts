import { useChangingHandle } from "@lbfalvy/react-utils";
import React from "react";
import ReactIs from "react-is";

/** Decides whether the component can be referenced */
function isRefable(comp: React.ComponentType<unknown>): boolean {
    return typeof comp !== 'function' // If it's a class component
        // or if it's one of these builtin components, which forward
        // refs.
        || ReactIs.isForwardRef(comp)
        || ReactIs.isLazy(comp)
        || ReactIs.isMemo(comp)
}

/**
 * Add entries to a ref such that `instanceof` checks continue to work,
 * `keyof` checks work if and only if the ref has no prototype, and
 * the handle never overrides indices that the ref has already defined.
 */
function fuseHandles<
    T extends Record<string, any>,
    U extends Record<string, any>
>(ref: T | null | undefined, handle: U | undefined): T & U | undefined {
    // If the ref is not set, don't return anything
    if (handle == null || handle == undefined) return ref as T & U;
    if (ref == null || ref == undefined) {
        return handle as T & U | undefined
    }
    // If it's a plain object, make it a subtype of the handle.
    if (Object.getPrototypeOf(ref) === null) {
        return Object.setPrototypeOf(ref, handle)
    }
    // If it's a class instance, copy keys over
    const refObj = (ref as any as Record<string, unknown>);
    for (let key of Object.keys(handle)) {
        // Only copy a key if it isn't defined on the target and also
        // evaluates to undefined when queried through the prototype chain
        let isOwnProp = Object.getOwnPropertyDescriptor(ref, key) !== undefined
        let isDefined = refObj[key] !== undefined
        if (!isOwnProp && !isDefined)
            refObj[key] = handle[key]
    }
    return refObj as T & U
    // We're making the assumption that all refs that have state are
    // classes. This makes sense because the three common types of ref
    // are Node, ReactComponent and imperative handles which are meant
    // to be disposable objects that collect functions.
}

/**
 * HKT for components that take a component as their prop. It enables
 * you to expose a handle with custom functions for imperative actions.
 * Your handle's functions will be copied over to the forwarded reference.
 * 
 * Typing most components returned by this function would require HKT
 * support in Typescript, so it's usually easiest to use an `as any as T`
 * unsafe cast.
 */
export function forwardRefWithHandle<
    UpProps,
    Handle extends Record<string, any>
>(
    render: (
        props: UpProps,
        useForwardedRef: (
            comp: React.ComponentType<any>
        ) => React.Ref<any> | null,
        setHandle?: 
            | ((handle: Handle) => void)
            | undefined,
    ) => React.ReactElement | null
): React.ForwardRefExoticComponent<
    & React.PropsWithoutRef<UpProps> 
    & React.RefAttributes<Handle>
> {
    return React.forwardRef<Handle, UpProps>((props, ref) => {
        let innerRef = React.useRef<Record<string, any>>();
        let handle = React.useRef<Handle|undefined>();
        const changeRef = useChangingHandle(ref)
        const updateRef = React.useCallback(() => {
            if (!changeRef) return
            let fused = fuseHandles(innerRef.current, handle.current)
            if (fused != null && fused != undefined) changeRef(fused)
        }, [changeRef])
        // Decides whether the given component currently needs a ref
        const useForwardedRef = (comp: React.ComponentType<any>): React.Ref<any> => {
            if (!changeRef || !isRefable(comp)) return null
            return React.useCallback((inner: any) => {
                innerRef.current = inner
                updateRef()
            }, [updateRef])
        }
        // Update the public-facing patch-on functions
        // if a ref was provided at all
        const setHandle = changeRef ? (newHandle: Handle) => {
            if (handle.current === newHandle) return;
            handle.current = newHandle
            updateRef()
        } : undefined
        return render(props, useForwardedRef, setHandle)
    })
}