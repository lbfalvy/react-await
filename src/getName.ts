export function getName(props: any): string {
    if ('displayName' in props && typeof props.displayName == 'string')
        return props.displayName
    if (
        typeof props.Comp == 'function'
        || typeof props.Comp == 'object'
    ) {
        const name = props.Comp.displayName
            ?? props.Comp.name
            ?? props.Comp.default?.displayName
            ?? props.Comp.default?.name
        if (typeof name == 'string') return name
    }
    return '[Component]'
}