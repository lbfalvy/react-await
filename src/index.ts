export { Await, awaitContext } from './Await';
export { AwaitCore } from './AwaitCore';