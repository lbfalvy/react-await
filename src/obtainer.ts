export type Obtainer = {
  readonly fn: Function,
  readonly deps?: readonly unknown[] | undefined,
}

export function runObtainer(
  obtainer: Obtainer | undefined,
  decl: Obtainer
): [Obtainer, unknown] | undefined {
  const { fn, deps } = decl;
  const changed = deps === undefined
    ? (obtainer?.fn !== fn || deps !== undefined)
    : (
      obtainer?.deps?.length !== deps.length
      || obtainer.deps.some((el, i) => el !== deps[i])
    );
  return changed ? [{ fn, deps }, fn()] : undefined;
}

