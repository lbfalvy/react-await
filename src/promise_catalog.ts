export type SettledStatus = ["succ" | "fail", unknown];

type PromiseStatus =
  | ["pending", ((status: SettledStatus) => void)[]]
  | SettledStatus;

type PromiseRecord = {
  id: number,
  status: PromiseStatus,
}

const PROMISES = new WeakMap<Promise<unknown>, PromiseRecord>();

let next_id = 0;

function mk_record(status: PromiseStatus): PromiseRecord {
  return { id: next_id++, status };
}

export function catalogPromise(promise: Promise<unknown>) {
  if (PROMISES.has(promise)) return;
  PROMISES.set(promise, mk_record(["pending", []]));
  promise.then(res => {
    const prev = PROMISES.get(promise)!;
    PROMISES.set(promise, mk_record(["succ", res]));
    if (prev.status[0] === "pending") {
      prev.status[1].forEach(f => f(["succ", res]));
    }
  }, err => {
    const prev = PROMISES.get(promise)!;
    PROMISES.set(promise, mk_record(["fail", err]));
    if (prev.status[0] === "pending") {
      prev.status[1].forEach(f => f(["fail", err]));
    }
  });
}

export function promiseStatus(
  promise: Promise<unknown>
): "pending" | SettledStatus {
  const record = PROMISES.get(promise);
  if (record == undefined || record.status[0] == "pending") return "pending";
  else return record.status;
}

export function then(
  promise: Promise<unknown>,
  cb: (status: ["succ" | "fail", unknown]) => void
) {
  if (!PROMISES.has(promise)) {
    catalogPromise(promise);
  }
  const record = PROMISES.get(promise)!;
  if (record.status[0] == "pending") record.status[1].push(cb);
  else Promise.resolve(record.status).then(cb);
}

export function debugPrint(promise: Promise<unknown>): string {
  const record = PROMISES.get(promise);
  if (record === undefined) return `Promise`;
  const { id, status: [variant, payload]} = record;
  if (variant == "pending") return `Promise(${id})`;
  return `Promise(${id}, ${variant}: ${payload})`;
}
