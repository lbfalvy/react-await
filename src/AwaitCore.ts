import { Obtainer, runObtainer } from "./obtainer";
import { SettledStatus, promiseStatus, then } from "./promise_catalog";
import * as promiseCatalog from "./promise_catalog";
import { Emit, Subscribe, event } from "@lbfalvy/mini-events";

export type PropDecl = {
  readonly obtainer?: Obtainer | undefined,
  readonly value?: unknown
};

export type AwaitStatus = "ready" | "pending" | "error";

export type StatusChange =
  | ["ready", Record<string, unknown>]
  | ["errors", Map<string, unknown>]

export class AwaitCore {
  private obtainers: Map<string, Obtainer> = new Map();
  private pendingCache: Map<string, Promise<unknown>> = new Map();
  private promiseCache: Map<Promise<unknown>, Set<string>> = new Map();
  private propsCache: Record<string, unknown> = {};
  private errorCache: Map<string, unknown> = new Map();
  private keys: Set<string> = new Set();
  public statusChange: Subscribe<[StatusChange]>;
  private emitStatusChange: Emit<[StatusChange]>;

  public constructor(
    props: Iterable<[string, PropDecl]>,
  ) {
    [this.emitStatusChange, this.statusChange] = event();
    for (const [key, prop] of props) {
      this.keys.add(key);
      if (prop.obtainer !== undefined) {
        const [obtainer, value] = runObtainer(undefined, prop.obtainer)!;
        this.obtainers.set(key, obtainer);
        this.handleMaybePromise(key, value);
      } else this.handleMaybePromise(key, prop.value);
    }
  }

  public update(props: Iterable<[string, PropDecl]>) {
    const old_keys = this.keys;
    this.keys = new Set();
    for (const [key, { obtainer, value }] of props) {
      this.keys.add(key);
      if (obtainer !== undefined) {
        const prev = this.obtainers.get(key);
        this.handleObtainer(key, prev, obtainer);
      } else {
        this.obtainers.delete(key);
        this.handleMaybePromise(key, value);
      }
    }
    for (const key of old_keys) {
      if (!this.keys.has(key)) {
        this.errorCache.delete(key);
        this.obtainers.delete(key);
        this.unsetPromise(key);
        delete this.propsCache[key];
      }
    }
  }

  public status(): AwaitStatus {
    if (0 < this.errorCache.size) return "error";
    if (0 < this.promiseCache.size) return "pending";
    return "ready";
  }

  public errors(): Map<string, unknown> | undefined {
    if (this.status() !== "error") return undefined;
    return new Map(this.errorCache);
  }

  public props(): Record<string, unknown> {
    return Object.assign({}, this.propsCache);
  }

  public reload() {
    for (const [key, obtainer] of this.obtainers) {
      this.handleObtainer(key, undefined, obtainer);
    }
  }

  private handleObtainer(key: string, prev: Obtainer|undefined, next: Obtainer) {
    const result = runObtainer(prev, next);
    if (result === undefined && prev === undefined) {
      throw new Error(
        "runObtainer returned undefined despite "
        + "previous obtainer being undefined"
      );
    }
    if (result === undefined) return;
    const [obtainer, value] = result;
    this.obtainers.set(key, obtainer);
    this.handleMaybePromise(key, value);
  }

  private handleMaybePromise(key: string, value: unknown) {
    if (value instanceof Promise) return this.handlePromise(key, value);
    this.unsetPromise(key);
    this.propsCache[key] = value;
  }

  private handlePromise(key: string, promise: Promise<unknown>) {
    const prevPromise = this.pendingCache.get(key);
    if (prevPromise === promise) return;
    const status = promiseStatus(promise);
    this.setPromise(key, promise);
    if (status == "pending") {
      then(promise, status => {
        this.onPromiseSettle(promise, status, this.emitStatusChange)
      })
    } else this.onPromiseSettle(promise, status, _ => { });
  }

  /// Returns true if there were previously no errors
  private handleError = (key: string, error: unknown) => {
    this.unsetPromise(key);
    this.errorCache.set(key, error);
  }

  /// Returns true if an error status has just been cleared
  private handleSuccess = (key: string, result: unknown) => {
    this.unsetPromise(key);
    this.errorCache.delete(key);
    this.propsCache[key] = result;
  }

  private setPromise(key: string, promise: Promise<unknown>) {
    this.unsetPromise(key);
    this.pendingCache.set(key, promise);
    const set = this.promiseCache.get(promise);
    if (set !== undefined) set.add(key);
    else this.promiseCache.set(promise, new Set([key]));
  }

  private unsetPromise(key: string): boolean {
    const pending = this.pendingCache.get(key);
    if (pending === undefined) return false;
    const aliases = this.promiseCache.get(pending)!;
    if (aliases?.size <= 1) this.promiseCache.delete(pending);
    else aliases.delete(key);
    this.pendingCache.delete(key);
    return true;
  }

  private onPromiseSettle(
    promise: Promise<unknown>,
    status: SettledStatus,
    onStatusChange: (status: StatusChange) => void
  ) {
    const [variant, payload] = status;
    const updated_keys = this.promiseCache.get(promise);
    if (updated_keys === undefined) return;
    if (updated_keys.size === 0)
      throw new Error("Empty promiseCache entry should have been removed");
    const prev_status = this.status();
    const handle_inst = {
      fail: this.handleError,
      succ: this.handleSuccess
    }[variant];
    for (const key of updated_keys) handle_inst(key, payload);
    if (this.status() !== prev_status) {
      if (this.status() === "ready")
        onStatusChange(["ready", this.props()!])
      else if (this.status() === "error")
        onStatusChange(["errors", this.errors()!]);
    }
  }

  public debugPrint(): string {
    return `AwaitCore{
      pendingCache: ${[...this.pendingCache].map(([k, pr]) =>
      `${k} => ${promiseCatalog.debugPrint(pr)}`
    )}
      promiseCache: ${[...this.promiseCache].map(([pr, kv]) =>
      `${promiseCatalog.debugPrint(pr)} => ${[...kv]}`
    )}
    }`;
  }
}
