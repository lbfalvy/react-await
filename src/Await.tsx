import React from 'react'
import { forwardRefWithHandle } from './forwardHandle'
import { getName } from './getName'
import { AwaitCore } from './AwaitCore'
import { ExtObtainer, MaybePromise, parseProps, WithPromisesAndObtainers } from './props'

export type Comp<T> =
    | React.ComponentType<T>
    | { default: React.ComponentType<T> }

export type ErrorReporter = (errors: Map<string, unknown>) => React.ReactNode;

// The props to be passed.
export type AwaitProps<T> =
    & ({
        Comp: MaybePromise<Comp<T>>
        f$Comp?: never
    } | {
        f$Comp: ExtObtainer<MaybePromise<Comp<T>>>
        Comp?: never
    })
    // Transform the props except children
    // reload is forced optional if present
    & WithPromisesAndObtainers<
        & Omit<T, 'children' | 'reload'>
        & ('reload' extends keyof T ? { reload?: T['reload'] } : {})
    >
    // Children is a special case to avoid nested XML tags if it isn't obtained.
    & ('children' extends keyof T ? (
        | {
            children?: MaybePromise<T['children']>
                | {
                    with?: MaybePromise<T['children']>
                    loader?: React.ReactNode
                    error?: ErrorReporter
                }
        }
        | {
            f$children: ExtObtainer<MaybePromise<T['children']>>
            children?: T['children'] & {
                loader?: React.ReactNode
                error?: ErrorReporter
            }
        }
    ) : {
        children?: {
            loader?: React.ReactNode
            error?: ErrorReporter
        }
    })

export type AwaitRef<T> =
    | { reload: () => void }
    | T & { reload: () => void }

/**
 * Deal with promises in React the easy way
 */
export const Await = forwardRefWithHandle(function Await(
    originalProps: AwaitProps<any>, useForwardedRef, setHandle
) {
    // extract explicit loader and error from children if present
    const awaitable = {...originalProps as any}
    let loader: React.ReactNode | undefined
    let error: ErrorReporter | undefined
    if ('children' in awaitable && typeof awaitable.children == 'object') {
        if ('with' in awaitable.children) {
            loader = awaitable.children.loader
            error = awaitable.children.error
            awaitable.children = awaitable.children.with
        }
    }
    // construct core on first run, update on subsequent runs
    const cRef = React.useRef<AwaitCore>();
    if (cRef.current) cRef.current.update(parseProps(awaitable));
    else cRef.current = new AwaitCore(parseProps(awaitable));
    const core = cRef.current; // this never changes
    const [status, setStatus] = React.useState(core.status());
    // subscribe to state changes once
    React.useLayoutEffect(() => {
        return core.statusChange(() => {
            // console.log("The state change got this far");
            setStatus(core.status())
        }, true)
    }, [])
    // // await all remaining props
    // const [allAvailable, status, reload] = useObtainAwait(awaitable)
    // initialise ref
    const [_, forceRerender] = React.useState({});
    const reload = React.useCallback(() => {
        core.reload();
        forceRerender({});
    }, []);
    setHandle?.({ reload });
    // remove component from props
    const { Comp, ...props } = core.props() as { Comp: any }
    // if component is a module, use its default export, else assume it
    // to be a component
    const Component =
        typeof Comp == 'object' // not a function component
        && !(Comp instanceof React.Component) // nor class component
        && 'default' in Comp // has a default member
        ? Comp.default as React.ComponentType<any>
        : Comp as React.ComponentType<any>
    // get forwarded ref
    const ref = useForwardedRef(Component)
    // default loader or error
    const ctx = React.useContext(awaitContext)

    // ######## EARLY RETURNS, NO HOOKS PAST THIS POINT ########
    if (status === "ready") {
        return <Component ref={ref} reload={reload} {...props} />
    }
    const name = getName(core.props());
    if (status == "error") {
        const errors = core.errors()!;
        if (error) return <>{error(errors)}</>
        return <ctx.error name={name} errors={errors} {...props} />
    }
    if (status == 'pending') {
        if (loader) return <>{loader}</>
        return <ctx.loader name={name} {...props} />
    }
    throw new Error("Invalid core status! (not ready/pending/error)");
}) as any as <T, Ref = {}>(
    props: AwaitProps<T> & {
        ref?: React.Ref<
            AwaitRef<Ref> | null | undefined
        > | null | undefined
    }
) => React.ReactElement

export const awaitContext = React.createContext({
    loader: ({ name }: { name: string }) => 
    <div className='await await-loading'>
        Loading {name}...
    </div>,
    error: ({ name, errors }: { name: string, errors: Map<string, unknown> }) => 
    <div className='await await-failed'>
        Failed to load {name}. Errors:
        <dl>
            {[...errors].map(([key, error]) => <div key={key}>
                <dt>{key}</dt>
                <dd>{`${error}`}</dd>
            </div>)}
        </dl>
    </div>
})