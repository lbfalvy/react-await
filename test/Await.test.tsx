import { when, XPromise } from '@lbfalvy/when';
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { Await } from '../src/Await';
import { Children, flushMtq, ShowString, TestWithReload } from './utils';

test('only given properties', () => {
  const component = renderer.create(
    <Await Comp={ShowString} s='foo' />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

describe('promise', () => {
  let p: XPromise<string>, res: (s: string) => void, rej: (e: any) => void
  let component: renderer.ReactTestRenderer
  beforeEach(() => {
    [p, res, rej] = when<string>()
    component = renderer.create(
      <Await Comp={ShowString} s={p} />
    )
  })

  test('resolves', async () => {
    await act(async () => {
      res('bar');
      await flushMtq();
    });
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('rejects', async () => {
    await act(async () => {
      rej('bar');
      await flushMtq();
    });
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('obtainer', () => {
  describe('synchronous', () => {
    test('without dependencies', () => {
      const component = renderer.create(<Await Comp={ShowString} f$s={() => 'foo'} />)
      expect(component.toJSON()).toMatchSnapshot()
    })
    
    describe('with dependencies', () => {
      let component: renderer.ReactTestRenderer
      let obtainS: jest.Mock<string, any>
      const idleProps = { Comp: ShowString };
      beforeEach(() => {
        obtainS = jest.fn().mockReturnValue('foo')
        component = renderer.create(<Await {...idleProps} f$s={[obtainS, 1]} />)
      })
      
      test('rerenders on dep change', () => {
        expect(obtainS).toHaveBeenCalledTimes(1)
        act(() => component.update(<Await {...idleProps} f$s={[obtainS, 2]} />))
        expect(obtainS).toHaveBeenCalledTimes(2)
      })
      
      test('does not rerender if only function changes', () => {
        const cb = jest.fn().mockReturnValue('quz')
        act(() => component.update(<Await {...idleProps} f$s={[cb, 1]} />))
        expect(cb).not.toHaveBeenCalled()
      })
    })
  })
  
  test('asynchronous', async () => {
    const promises: [XPromise<string>, (s: string) => void, (e: any) => void][] = []
    const obtainer = () => {
      const [p, rs, rj] = when<string>()
      promises.push([p, rs, rj])
      return p
    }
    const component = renderer.create(<Await Comp={ShowString} f$s={obtainer} />)
    expect(promises).toHaveLength(1)
    await act(async () => {
      promises[0][1]('foo');
      await flushMtq();
    })
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('children', () => {
  test('plain', () => {
    const component = renderer.create(<Await Comp={Children} s='foo'>bar</Await>)
    expect(component.toJSON()).toMatchSnapshot()
  })
  
  test('with custom loader', async () => {
    const [promise, resolve, reject] = when<string>()
    const component = renderer.create(<Await Comp={Children} s={promise}>{{
      with: 'bar',
      loader: 'waiting for data...'
    }}</Await>)
    expect(component.toJSON()).toMatchSnapshot()
    await act(async () => {
      resolve('foo');
      await flushMtq();
    })
    expect(component.toJSON()).toMatchSnapshot()
  })
  
  test('promise', async () => {
    const [promise, resolve] = when<string>()
    const component = renderer.create(<Await Comp={Children} s='foo'>{promise}</Await>)
    expect(component.toJSON()).toMatchSnapshot()
    await act(async () => {
      resolve('bar');
      await flushMtq();
    })
    expect(component.toJSON()).toMatchSnapshot()
  })
  
  test('obtained', () => {
    const component = renderer.create(<Await Comp={Children} s='foo' f$children={[() => 'bar']} />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
  
describe('reload', () => {
  test('externally', () => {
    let ref: { reload: () => void } | null | undefined
    const component = renderer.create(<Await
      Comp={TestWithReload}
      ref={r => ref = r}
      f$s={[() => "foo"]}
    />)
    act(() => ref?.reload())
    expect(component.toJSON()).toMatchSnapshot()
  })

  test('internally', () => {
    let reload!: () => void;
    const component = renderer.create(<Await
      Comp={TestWithReload}
      withReload={(r: () => void) => reload = r}
      f$s={[() => "foo"]}
    />)
    act(() => reload())
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('Async obtainer for component', () => {
  test('plain', async () => {
    const [promise, resolve, reject] = when<(props: { s: string }) => React.ReactElement>()
    const component = renderer.create(<Await f$Comp={[() => promise]} s='foo' />)
    await act(async () => {
      resolve(ShowString);
      await flushMtq();
    })
    expect(component.toJSON()).toMatchSnapshot()
  })
  test('in module', async () => {
    const [promise, resolve, reject] = when<{ default: (props: { s: string }) => React.ReactElement }>()
    const component = renderer.create(<Await f$Comp={[() => promise]} s='foo' />)
    await act(async () => {
      resolve({ default: ShowString })
      await flushMtq();
    })
    expect(component.toJSON()).toMatchSnapshot()
  })
})