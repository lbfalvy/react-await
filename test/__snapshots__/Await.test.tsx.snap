// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async obtainer for component in module 1`] = `"foo"`;

exports[`Async obtainer for component plain 1`] = `"foo"`;

exports[`children obtained 1`] = `
[
  "foo",
  ";",
  "bar",
]
`;

exports[`children plain 1`] = `
[
  "foo",
  ";",
  "bar",
]
`;

exports[`children promise 1`] = `
<div
  className="await await-loading"
>
  Loading 
  Children
  ...
</div>
`;

exports[`children promise 2`] = `
[
  "foo",
  ";",
  "bar",
]
`;

exports[`children with custom loader 1`] = `"waiting for data..."`;

exports[`children with custom loader 2`] = `
[
  "foo",
  ";",
  "bar",
]
`;

exports[`obtainer asynchronous 1`] = `"foo"`;

exports[`obtainer synchronous without dependencies 1`] = `"foo"`;

exports[`only given properties 1`] = `"foo"`;

exports[`promise rejects 1`] = `
<div
  className="await await-failed"
>
  Failed to load 
  ShowString
  . Errors:
  <dl>
    <div>
      <dt>
        s
      </dt>
      <dd>
        bar
      </dd>
    </div>
  </dl>
</div>
`;

exports[`promise resolves 1`] = `"bar"`;

exports[`reload externally 1`] = `"2"`;

exports[`reload internally 1`] = `"2"`;
